ujt/callbacks.py

```diff
@@ -40,54 +32,137 @@
+        user_journey_table_selected_row_ids: List of selected row ids from the user journey datatable. Should contain only one element. Used for highlighting a path through the graph.  
+        virtual_node_update_signal: String used as a signal to indicate that the virtual node addition/deletion was valid. 
+        collapse_n_clicks_timestamp: Timestamp of when the collapse button was clicked. Value unused, input only provided to register callback.
+        expand_n_clicks_timestamp: Timestamp of when the expand button was clicked. Value unused, input only provided to register callback.
+        state_elements: The list of current cytoscape graph elements. This is unused and can be removed in a later change. 
+        selected_node_data: The list of data dictionaries for selected nodes. Used to create virtual nodes.
+        virtual_node_input_value: The value of the virtual node input box. Used to perform all virtual node operations. 
     Returns:
         A dictionary of cytoscape elements describing the nodes and edges of the graph.
     """
 
-    # regenerate elements on initial call and callbacks invoked by refresh button
-    regenerate_elements = True
-
     ctx = dash.callback_context
+    triggered_id, triggered_prop, triggered_value = utils.ctx_triggered_info(ctx)
 
-    if ctx.triggered:
-        # ctx.triggered[0] is either "cytoscape-graph.tapNode" or "client-dropdown.value"
-        triggered_id, triggered_prop = ctx.triggered[0]["prop_id"].split(".")
-        if triggered_id == "refresh-button":
-            state.clear_cache()
-        else:
-            # don't regenerate elements when triggered by a client datatable row selection
-            regenerate_elements = False
-
-    if regenerate_elements:
-        node_name_message_map, client_name_message_map = state.get_message_maps()
+    if triggered_id == "virtual-node-update-signal" and triggered_value != constants.OK_SIGNAL:
+        # No-op if the validation signal isn't OK
+        raise PreventUpdate
 
-        cytoscape_graph_elements = converters.cytoscape_elements_from_maps(
-            node_name_message_map,
-            client_name_message_map,
-        )
+    if triggered_id == "refresh-button":
+        state.get_slis()  # no-op for now.
+
+    # Perform status computation.
+    node_name_message_map, client_name_message_map = state.get_message_maps()
+    virtual_node_map = state.get_virtual_node_map()
+
+    compute_status.reset_node_statuses(node_name_message_map)
+    compute_status.reset_client_statses(client_name_message_map)
+    compute_status.reset_node_statuses(virtual_node_map)
+
+    # combine the two maps of nodes into one dictionary
+    # use duck typing -- is this pythonic or a hack?
+    all_nodes_map = {**node_name_message_map, **virtual_node_map}  #type: ignore
+    compute_status.compute_statuses(
+        all_nodes_map,
+        client_name_message_map,
+    )
```

Comment:
Interested to hear your thoughts about this.
Virtual Nodes have children and their status should be computed based on the status of their children.
However, virtual nodes don't have dependencies. I take advantage of python's dick typing here to reuse the existing status computation code for virtual nodes. I think this approach is more elegant, but can also see why we would want to have a specific function to compute status for virtual nodes, even if some of the code is repeated.

(The same situation arises in the callback to create the node info panel.)

Reply:
this is probably better than having two separate functions.